### PR TITLE
feat!(runner): support concurrent suites

### PR DESCRIPTION
**Pull request by hi-ogawa**
### Description

- closes https://github.com/vitest-dev/vitest/issues/5458

PoC of running multiple suites concurrently.

In terms of implementation, it looks like it's a matter of setting `Suite.concurrent = true` in some way. My current approach is to have an explicit flag `TestOptions.concurrentSuite: boolean` and the idea is based on the analogy:
- `describe("...", { concurrentSuite: true }, ...)` to run neighboring suites concurrently
- `test("...", { concurrent: true }, ...)` to run neighboring tests concurrently

TBD 
- ~comparison with Jest?~
  - they runs all `test.concurrent` first whether it is inside `describe` and they have `before/after` hooks related bugs.
- breaking change? (change `describe.concurrent` to affect all suite/task level concurrency)
- limit concurrency only for tests but not for suites? otherwise, it's easy to make deadlock where outer concurrent `describe` is waiting for inner concurrent `test` to finish, then inner one is blocked by `p-limit`.

TODO
- [x] doc

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by jgoux** [comment]
Hey @hi-ogawa, thanks for implementing it.

I'm wondering why "concurrentSuite" is needed? Shouldn't we mimic jest and have all the describe blocks concurrent? Or is it for backward compat only?

If people like me want jest behaviour on all describe blocks, could we add a concurrentSuite parameter in vitest.config.ts so it's enabled globally?

**Comment by hi-ogawa** [comment]
> I'm wondering why "concurrentSuite" is needed? Shouldn't we mimic jest and have all the describe blocks concurrent? Or is it for backward compat only?

I think I was mostly concerned with back compat, but anyways this PR is just for a starting point of discussion to explore the idea. Depending on what users expect and how Jest has been doing, we'll discuss further what this feature should be.

File: packages/runner/src/types/tasks.ts [hunk 1]
```diff
@@ -185,6 +185,10 @@ export interface TestOptions {
    * Tests inherit `concurrent` from `describe()` and nested `describe()` will inherit from parent's `concurrent`.
    */
   concurrent?: boolean
+  /**
+   * TODO: suite only options?
+   */
+  concurrentSuite?: boolean
```
Comment (sheremet-va) [inline packages/runner/src/types/tasks.ts:191]: I don't really like having a separate option. Do we really need it?

Comment (hi-ogawa) [reply packages/runner/src/types/tasks.ts:191]: Oh, forgot about this TODO.
If we don't differentiate suite and task level concurrency, then single flag should suffice.
let me check it again.

Comment (hi-ogawa) [reply packages/runner/src/types/tasks.ts]: Actually I wasn't so sure about the breaking change, but this behavior probably makes sense.
Users can still make inner suite/tests sequential easily via `describe.sequential` just like described in the doc https://vitest.dev/api/#describe-sequential

File: test/cli/fixtures/fails/concurrent-suite-deadlock.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { describe, test, vi } from 'vitest'
+import { createDefer } from 'vitest/dist/utils.js'
```
Comment (sheremet-va) [inline test/cli/fixtures/fails/concurrent-suite-deadlock.test.ts:2]: Why is it importing from dist 🤔

Comment (hi-ogawa) [reply test/cli/fixtures/fails/concurrent-suite-deadlock.test.ts]: Hmm, I didn't notice but it seems typescript got that one from autocomplete.
I installed `@vitest/utils` and it looks okay now.

**Review by sheremet-va — APPROVED** [review]

**Comment by sheremet-va** [comment]
> * they runs all `test.concurrent` first whether it is inside `describe` and they have `before/after` hooks related bugs.
> * breaking change? (change `describe.concurrent` to affect all suite/task level concurrency)
> * limit concurrency only for tests but not for suites? otherwise, it's easy to make deadlock where outer concurrent `describe` is waiting for inner concurrent `test` to finish, then inner one is blocked by `p-limit`.

Actually, were these points addressed in any way? 😄

**Pull request merged**